# What drives the price of a car?

![](images/kurt.jpeg)

**OVERVIEW**

In this application, you will explore a dataset from kaggle. The original dataset contained information on 3 million used cars. The provided dataset contains information on 426K cars to ensure speed of processing.  Your goal is to understand what factors make a car more or less expensive.  As a result of your analysis, you should provide clear recommendations to your client -- a used car dealership -- as to what consumers value in a used car.

In [1]:
#INSTALLAITONS
from sklearn.linear_model import Ridge, Lasso, LinearRegression
from sklearn.feature_selection import SequentialFeatureSelector, RFE
from sklearn.pipeline import Pipeline

from sklearn.linear_model import LinearRegression, Lasso
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.feature_selection import SequentialFeatureSelector, SelectFromModel
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn import set_config
set_config(display="diagram")


import pandas as pd
import numpy as np
import seaborn as sns
import plotly.express as px

### CRISP-DM Framework

<center>
    <img src = images/crisp.png width = 50%/>
</center>


To frame the task, throughout our practical applications we will refer back to a standard process in industry for data projects called CRISP-DM.  This process provides a framework for working through a data problem.  Your first step in this application will be to read through a brief overview of CRISP-DM [here](https://mo-pcco.s3.us-east-1.amazonaws.com/BH-PCMLAI/module_11/readings_starter.zip).  After reading the overview, answer the questions below.

### Business Understanding

From a business perspective, we are tasked with identifying key drivers for used car prices.  In the CRISP-DM overview, we are asked to convert this business framing to a data problem definition.  Using a few sentences, reframe the task as a data task with the appropriate technical vocabulary. 

Step 1 
- business objectives: we wish to know what makes a car more or less expensive


Step 2 
- Assess situation: our current situation is a data set of 400'000 rows


Step 3 
- data mining goals: Predict the valuation on a car based on the atributes of a car (make/model/year/location)

Step 4
- product plan: with minor details provided by a customer, we will provide a reliable estimate of price for the car


### Data Understanding

After considering the business understanding, we want to get familiar with our data.  Write down some steps that you would take to get to know the dataset and identify any quality issues within.  Take time to get to know the dataset and explore what information it contains and how this could be used to inform your business understanding.

In [2]:
# Step 1: collect initial data
df = pd.read_csv('./data/vehicles.csv')

In [3]:
# Step 2: describe the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   region        426880 non-null  object 
 2   price         426880 non-null  int64  
 3   year          425675 non-null  float64
 4   manufacturer  409234 non-null  object 
 5   model         421603 non-null  object 
 6   condition     252776 non-null  object 
 7   cylinders     249202 non-null  object 
 8   fuel          423867 non-null  object 
 9   odometer      422480 non-null  float64
 10  title_status  418638 non-null  object 
 11  transmission  424324 non-null  object 
 12  VIN           265838 non-null  object 
 13  drive         296313 non-null  object 
 14  size          120519 non-null  object 
 15  type          334022 non-null  object 
 16  paint_color   296677 non-null  object 
 17  state         426880 non-null  object 
dtypes: f

In [4]:
# Step 3: explore the data
df.head()

# Summary statistics
print(df.describe())

# Data types and missing values
print(df.info())

                 id         price           year      odometer
count  4.268800e+05  4.268800e+05  425675.000000  4.224800e+05
mean   7.311487e+09  7.519903e+04    2011.235191  9.804333e+04
std    4.473170e+06  1.218228e+07       9.452120  2.138815e+05
min    7.207408e+09  0.000000e+00    1900.000000  0.000000e+00
25%    7.308143e+09  5.900000e+03    2008.000000  3.770400e+04
50%    7.312621e+09  1.395000e+04    2013.000000  8.554800e+04
75%    7.315254e+09  2.648575e+04    2017.000000  1.335425e+05
max    7.317101e+09  3.736929e+09    2022.000000  1.000000e+07
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   region        426880 non-null  object 
 2   price         426880 non-null  int64  
 3   year          425675 non-null  float64
 4   manufacturer  409234 non-null  object 
 5   model    

In [5]:
df['type'].unique()
#df['condition'].unique()

array([nan, 'pickup', 'truck', 'other', 'coupe', 'SUV', 'hatchback',
       'mini-van', 'sedan', 'offroad', 'bus', 'van', 'convertible',
       'wagon'], dtype=object)

In [6]:
# # Convert categorical variable to numerical using one-hot encoding
# from matplotlib import pyplot as plt
# df = df.drop(columns=['VIN'])
# data_encoded = pd.get_dummies(df, columns=['region','manufacturer', 'model', 'condition', 'cylinders', 'fuel', 'title_status', 'paint_color', 'type', 'size', 'drive', 'transmission', 'title_status', 'state'])

# # Calculate correlation matrix
# corr_matrix = data_encoded.corr()

# # Plot correlation matrix as heatmap
# plt.figure(figsize=(8, 6))
# sns.heatmap(corr_matrix, annot=True, cmap='coolwarm', fmt=".2f")
# plt.title('Correlation Matrix')
# plt.show()

In [7]:
# Step 4: verify data quality
# missing values in 14/18 of the columns
# will need to handle some of the 'object' data types

# this will require me to convert these object types into unique collumns and attribute them as boolean variables
# Will need to handle the missing values by either filling in them with expected value or drop them

### Data Preparation

After our initial exploration and fine tuning of the business understanding, it is time to construct our final dataset prior to modeling.  Here, we want to make sure to handle any integrity issues and cleaning, the engineering of new features, any transformations that we believe should happen (scaling, logarithms, normalization, etc.), and general preparation for modeling with `sklearn`. 

In [8]:
# Step 1 - Select the data
df

id                  region  price    year manufacturer  \
0       7222695916                prescott   6000     NaN          NaN   
1       7218891961            fayetteville  11900     NaN          NaN   
2       7221797935            florida keys  21000     NaN          NaN   
3       7222270760  worcester / central MA   1500     NaN          NaN   
4       7210384030              greensboro   4900     NaN          NaN   
...            ...                     ...    ...     ...          ...   
426875  7301591192                 wyoming  23590  2019.0       nissan   
426876  7301591187                 wyoming  30590  2020.0        volvo   
426877  7301591147                 wyoming  34990  2020.0     cadillac   
426878  7301591140                 wyoming  28990  2018.0        lexus   
426879  7301591129                 wyoming  30590  2019.0          bmw   

                           model condition    cylinders    fuel  odometer  \
0                            NaN       NaN          NaN     NaN       NaN   
1                            NaN       NaN          NaN     NaN       NaN   
2                            NaN       NaN          NaN     NaN       NaN   
3                            NaN       NaN          NaN     NaN       NaN   
4                            NaN       NaN          NaN     NaN       NaN   
...                          ...       ...          ...     ...       ...   
426875         maxima s sedan 4d      good  6 cylinders     gas   32226.0   
426876  s60 t5 momentum sedan 4d      good          NaN     gas   12029.0   
426877          xt4 sport suv 4d      good          NaN  diesel    4174.0   
426878           es 350 sedan 4d      good  6 cylinders     gas   30112.0   
426879  4 series 430i gran coupe      good          NaN     gas   22716.0   

       title_status transmission                VIN drive size       type  \
0               NaN          NaN                NaN   NaN  NaN        NaN   
1               NaN          NaN                NaN   NaN  NaN        NaN   
2               NaN          NaN                NaN   NaN  NaN        NaN   
3               NaN          NaN                NaN   NaN  NaN        NaN   
4               NaN          NaN                NaN   NaN  NaN        NaN   
...             ...          ...                ...   ...  ...        ...   
426875        clean        other  1N4AA6AV6KC367801   fwd  NaN      sedan   
426876        clean        other  7JR102FKXLG042696   fwd  NaN      sedan   
426877        clean        other  1GYFZFR46LF088296   NaN  NaN  hatchback   
426878        clean        other  58ABK1GG4JU103853   fwd  NaN      sedan   
426879        clean        other  WBA4J1C58KBM14708   rwd  NaN      coupe   

       paint_color state  
0              NaN    az  
1              NaN    ar  
2              NaN    fl  
3              NaN    ma  
4              NaN    nc  
...            ...   ...  
426875         NaN    wy  
426876         red    wy  
426877       white    wy  
426878      silver    wy  
426879         NaN    wy  

[426880 rows x 18 columns]

In [9]:
# Step 2 - Clean the data
# Drop rows with missing values for columns with a small number of missing values
df.dropna(subset=['year', 'manufacturer', 'model', 'fuel', 'odometer', 'title_status', 'transmission'], inplace=True)

# Fill missing values for columns with a large number of missing values
df['condition'].fillna(value='unknown', inplace=True)
df['cylinders'].fillna(value='unknown', inplace=True)
df['VIN'].fillna(value='unknown', inplace=True)
df['drive'].fillna(value='unknown', inplace=True)
df['size'].fillna(value='unknown', inplace=True)
df['type'].fillna(value='unknown', inplace=True)
df['paint_color'].fillna(value='unknown', inplace=True)

# Convert data types
df['year'] = df['year'].astype(int)
df['odometer'] = df['odometer'].astype(int)

# Remove duplicate rows
df.drop_duplicates(inplace=True)

# Check for and handle outliers if necessary
# Save the cleaned dataset
# df.to_csv('cleaned_dataset.csv', index=False)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 389604 entries, 27 to 426879
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   id            389604 non-null  int64 
 1   region        389604 non-null  object
 2   price         389604 non-null  int64 
 3   year          389604 non-null  int64 
 4   manufacturer  389604 non-null  object
 5   model         389604 non-null  object
 6   condition     389604 non-null  object
 7   cylinders     389604 non-null  object
 8   fuel          389604 non-null  object
 9   odometer      389604 non-null  int64 
 10  title_status  389604 non-null  object
 11  transmission  389604 non-null  object
 12  VIN           389604 non-null  object
 13  drive         389604 non-null  object
 14  size          389604 non-null  object
 15  type          389604 non-null  object
 16  paint_color   389604 non-null  object
 17  state         389604 non-null  object
dtypes: int64(4), object(14)
memo

In [11]:
# Step 3 - Construct the data
# Define mapping of condition values to integers
condition_mapping = {
    'salvage': 0,
    'poor': 1,
    'fair': 2,
    'good': 3,
    'like new': 4,
    'excellent': 5,
    'new': 6,
}

# Map condition values to integers
df['condition'] = df['condition'].map(condition_mapping)

# Define a function to extract the number of cylinders
def extract_cylinders(s):
    if pd.isnull(s) or s == 'unknown' or s == 'other':
        return 0
    else:
        return int(s.split()[0])

# Apply the function to the 'cylinders' column
df['cylinders'] = df['cylinders'].apply(extract_cylinders)

#Dropping VIN as it wont impact valuation
df = df.drop(columns=['VIN'])




In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 389604 entries, 27 to 426879
Data columns (total 17 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            389604 non-null  int64  
 1   region        389604 non-null  object 
 2   price         389604 non-null  int64  
 3   year          389604 non-null  int64  
 4   manufacturer  389604 non-null  object 
 5   model         389604 non-null  object 
 6   condition     232322 non-null  float64
 7   cylinders     389604 non-null  int64  
 8   fuel          389604 non-null  object 
 9   odometer      389604 non-null  int64  
 10  title_status  389604 non-null  object 
 11  transmission  389604 non-null  object 
 12  drive         389604 non-null  object 
 13  size          389604 non-null  object 
 14  type          389604 non-null  object 
 15  paint_color   389604 non-null  object 
 16  state         389604 non-null  object 
dtypes: float64(1), int64(5), object(11)
memory usage: 53

In [13]:
#df.fillna('unknown', inplace=True)  # Fill other categorical columns with 'unknown'
from sklearn.calibration import LabelEncoder


df['condition'].fillna(df['condition'].median(), inplace=True)

# Convert data types
df['condition'] = df['condition'].astype(int)
df['cylinders'] = df['cylinders'].astype(int)
# Fit and transform the 'model' column
label_encoder = LabelEncoder()

df['model_encoded'] = label_encoder.fit_transform(df['model'].astype(str))

# Fit and transform the 'manufacturer' column
df['manufacturer_encoded'] = label_encoder.fit_transform(df['manufacturer'].astype(str))

df['region_encoded'] = label_encoder.fit_transform(df['region'].astype(str))

df['state_encoded'] = label_encoder.fit_transform(df['state'].astype(str))

df = df.drop(columns=['region'])
df = df.drop(columns=['state'])
df = df.drop(columns=['manufacturer'])
df = df.drop(columns=['model'])




# Categorical encoding (One-hot encoding)
cat_columns = ['fuel', 'title_status', 'transmission', 'drive', 'size', 'type', 'paint_color']
df = pd.get_dummies(df, columns=cat_columns)

# Feature engineering (Example: Calculate car age)
current_year = pd.Timestamp.now().year
df['car_age'] = current_year - df['year']

In [14]:
df

id  price  year  condition  cylinders  odometer  \
27      7316814884  33590  2014          3          8     57923   
28      7316814758  22590  2010          3          8     71229   
29      7316814989  39590  2020          3          8     19160   
30      7316743432  30990  2017          3          8     41124   
31      7316356412  15000  2013          5          6    128000   
...            ...    ...   ...        ...        ...       ...   
426875  7301591192  23590  2019          3          6     32226   
426876  7301591187  30590  2020          3          0     12029   
426877  7301591147  34990  2020          3          0      4174   
426878  7301591140  28990  2018          3          6     30112   
426879  7301591129  30590  2019          3          0     22716   

        model_encoded  manufacturer_encoded  region_encoded  state_encoded  \
27              17164                    14              16              1   
28              17546                     7              16              1   
29              17571                     7              16              1   
30              20272                    38              16              1   
31               8732                    13              16              1   
...               ...                   ...             ...            ...   
426875          13281                    30             397             50   
426876          16586                    40             397             50   
426877          21577                     6             397             50   
426878           7828                    23             397             50   
426879           1664                     4             397             50   

        ...  paint_color_green  paint_color_grey  paint_color_orange  \
27      ...              False             False               False   
28      ...              False             False               False   
29      ...              False             False               False   
30      ...              False             False               False   
31      ...              False             False               False   
...     ...                ...               ...                 ...   
426875  ...              False             False               False   
426876  ...              False             False               False   
426877  ...              False             False               False   
426878  ...              False             False               False   
426879  ...              False             False               False   

        paint_color_purple  paint_color_red  paint_color_silver  \
27                   False            False               False   
28                   False            False               False   
29                   False             True               False   
30                   False             True               False   
31                   False            False               False   
...                    ...              ...                 ...   
426875               False            False               False   
426876               False             True               False   
426877               False            False               False   
426878               False            False                True   
426879               False            False               False   

        paint_color_unknown  paint_color_white  paint_color_yellow  car_age  
27                    False               True               False       10  
28                    False              False               False       14  
29                    False              False               False        4  
30                    False              False               False        7  
31                    False              False               False       11  
...                     ...                ...                 ...      ...  
426875                 True              False            

In [15]:
# Step 4 - Integrate the data
#there is no plan to join tables so we will skip this step
#But below is sample code to conduct such activities



# # Load the data from multiple tables or records
# # Replace 'table1.csv', 'table2.csv', etc. with the actual file paths
# table1 = pd.read_csv('table1.csv')
# table2 = pd.read_csv('table2.csv')
# # Load additional tables if needed

# # Perform data integration
# # Example 1: Merge tables based on common key(s)
# merged_data = pd.merge(table1, table2, on='common_key', how='inner')

# # Example 2: Concatenate tables along rows (stack vertically)
# concatenated_data = pd.concat([table1, table2], axis=0)

# # Example 3: Concatenate tables along columns (stack horizontally)
# concatenated_data = pd.concat([table1, table2], axis=1)

# # Example 4: Join tables based on common key(s)
# # For more complex joins, you can use the merge function with different join types (inner, outer, left, right)

# # Example 5: Append one table to another
# appended_data = table1.append(table2)

# # Example 6: Combine tables using database-style join operations
# # You can use the merge function with different parameters to perform database-style joins

# # Example 7: Perform more complex data integration operations as needed

# # Display the integrated data
# print(merged_data.head())
# print(concatenated_data.head())
# print(appended_data.head())

In [16]:
# Step 5 - Format the data

# Perform formatting transformations
# Example 1: Convert column names to lowercase
#data.columns = data.columns.str.lower()
# not required

# Example 2: Remove leading and trailing whitespaces from column values
#data['column_name'] = data['column_name'].str.strip()
# we wont use this since they are all good

# Example 3: Convert categorical variables to numerical values using label encoding
# from sklearn.preprocessing import LabelEncoder
# encoder = LabelEncoder()
#data['encoded_column'] = encoder.fit_transform(data['categorical_column'])
# we wont use this as we already Hot Encoded it

# Example 4: Convert datetime strings to datetime objects
# data['date_column'] = pd.to_datetime(data['date_column'])
#no date time

# Example 5: Convert numerical values to categorical variables based on bins
# data['bin_column'] = pd.cut(data['numeric_column'], bins=5, labels=['bin1', 'bin2', 'bin3', 'bin4', 'bin5'])
# handled with hot encoding

# Example 6: Convert boolean values to integers
# data['boolean_column'] = data['boolean_column'].astype(int)
#already done 

# Example 7: Convert text data to lowercase or uppercase
# data['text_column'] = data['text_column'].str.lower()
# not required

# Example 8: Replace missing values with a specific value
# data.fillna(value='missing', inplace=True)
# already handled

# Example 9: Normalize numerical values to a specific range
# from sklearn.preprocessing import MinMaxScaler
# scaler = MinMaxScaler()
# data['normalized_column'] = scaler.fit_transform(data[['numerical_column']])
#scaling will be done later

# Example 10: Perform additional formatting transformations as needed

# Display the formatted data
print(df.info)
# Display additional formatted data as needed


<bound method DataFrame.info of                 id  price  year  condition  cylinders  odometer  \
27      7316814884  33590  2014          3          8     57923   
28      7316814758  22590  2010          3          8     71229   
29      7316814989  39590  2020          3          8     19160   
30      7316743432  30990  2017          3          8     41124   
31      7316356412  15000  2013          5          6    128000   
...            ...    ...   ...        ...        ...       ...   
426875  7301591192  23590  2019          3          6     32226   
426876  7301591187  30590  2020          3          0     12029   
426877  7301591147  34990  2020          3          0      4174   
426878  7301591140  28990  2018          3          6     30112   
426879  7301591129  30590  2019          3          0     22716   

        model_encoded  manufacturer_encoded  region_encoded  state_encoded  \
27              17164                    14              16              1   
28     

### Modeling

With your (almost?) final dataset in hand, it is now time to build some models.  Here, you should build a number of different regression models with the price as the target.  In building your models, you should explore different parameters and be sure to cross-validate your findings.

In [17]:
#4.1 Select modeling technique
# 3. Modeling
from sklearn.cluster import KMeans
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import silhouette_score, accuracy_score, r2_score, mean_squared_error

# Extract features (X) and target (y)
X = df.drop(columns=['price'])  
y = df['price']  # Select the 'price' column as the target variable

# Optionally, you can perform further preprocessing on X and y, such as handling missing values, encoding categorical variables, or scaling numeric features.

# Example preprocessing (replace with your actual preprocessing steps)
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

# # Handle missing values (replace NaNs with median)
# imputer = SimpleImputer(strategy='median')
# X = imputer.fit_transform(X)

# # Scale features to have zero mean and unit variance
# scaler = StandardScaler()
# X = scaler.fit_transform(X)

# Now, X and y are ready for modeling

# 4.2 Generate test design
# Split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [ ]:

# 4.3 Build model
# Clustering (K-Means)
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import numpy as np

# Define a range of cluster numbers to try
k_values = range(2, 11)  # Try clusters from 2 to 10

# Initialize lists to store scores and models
silhouette_scores = []
kmeans_models = []

# Iterate over each value of k
for k in k_values:
    # Create KMeans instance
    kmeans = KMeans(n_clusters=k, random_state=42)
    
    # Fit KMeans model to training data
    kmeans.fit(X_train)
    
    # Compute silhouette score
    silhouette_avg = silhouette_score(X_train, kmeans.labels_)
    
    # Append silhouette score and model to lists
    silhouette_scores.append(silhouette_avg)
    kmeans_models.append(kmeans)

# Find the index of the best silhouette score
best_index = np.argmax(silhouette_scores)

# Get the best KMeans model and corresponding silhouette score
best_kmeans_model = kmeans_models[best_index]
best_silhouette_score = silhouette_scores[best_index]

# Print the best silhouette score and number of clusters
print("Best Silhouette Score:", best_silhouette_score)
print("Number of Clusters for Best Silhouette Score:", k_values[best_index])

# Now you can use the best_kmeans_model for further analysis or clustering



/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


In [19]:

# Linear Regression
linear_reg = LinearRegression()
linear_reg.fit(X_train, y_train)


LinearRegression()

In [23]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV

# Define the parameter grid for Ridge regression
ridge_param_grid = {
    'alpha': [0.001, 0.01, 0.1, 1.0, 10.0]  # Regularization strength values to try
}

# Perform grid search for Ridge regression
ridge_grid_search = GridSearchCV(Ridge(), ridge_param_grid, cv=5)
ridge_grid_search.fit(X_train, y_train)

# Get the best hyperparameters and retrain the Ridge regression model
best_ridge_model = ridge_grid_search.best_estimator_
best_ridge_model.fit(X_train, y_train)

# Assess the performance of the best Ridge regression model
ridge_r2 = best_ridge_model.score(X_test, y_test)
ridge_rmse = mean_squared_error(y_test, best_ridge_model.predict(X_test), squared=False)

print("Best Ridge Regression - R-squared:", ridge_r2)
print("Best Ridge Regression - RMSE:", ridge_rmse)


/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:211: LinAlgWarning: Ill-conditioned matrix (rcond=2.09407e-22): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:211: LinAlgWarning: Ill-conditioned matrix (rcond=2.09649e-22): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:211: LinAlgWarning: Ill-conditioned matrix (rcond=2.09345e-22): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:211: LinAlgWarning: Ill-conditioned matrix (rcond=2.09442e-22): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/Users/mrburstein/an

Best Ridge Regression - R-squared: 5.603348547011944e-05
Best Ridge Regression - RMSE: 14865774.976668088


In [20]:
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier

# Define parameter grid for KNN classifier
knn_classifier_param_grid = {
    'n_neighbors': [3, 5, 7, 9],
    'weights': ['uniform', 'distance'],
    'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
}

# Perform grid search for KNN classifier
knn_classifier_grid_search = GridSearchCV(KNeighborsClassifier(), knn_classifier_param_grid, cv=5, n_jobs=-1)
knn_classifier_grid_search.fit(X_train, y_train)

# Get best hyperparameters and retrain KNN classifier
best_knn_classifier = knn_classifier_grid_search.best_estimator_
best_knn_classifier.fit(X_train, y_train)

# Evaluate the best KNN classifier on the test set
best_knn_classifier_accuracy = best_knn_classifier.score(X_test, y_test)

print("Best KNN Classifier Accuracy:", best_knn_classifier_accuracy)


/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/model_selection/_split.py:725: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:824: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/model_selection/_validation.py", line 813, in _score
    scores = scorer(estimator, X_test, y_test)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_scorer.py", line 527, in __call__
    return estimator.score(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/base.py", line 705, in score
    return 

Best KNN Classifier Accuracy: 0.19202782305155222


In [21]:
# Define parameter grid for KNN regressor
knn_regressor_param_grid = {
    'n_neighbors': [3, 5, 7, 9],
    'weights': ['uniform', 'distance'],
    'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
}

# Perform grid search for KNN regressor
knn_regressor_grid_search = GridSearchCV(KNeighborsRegressor(), knn_regressor_param_grid, cv=5, n_jobs=-1)
knn_regressor_grid_search.fit(X_train, y_train)

# Get best hyperparameters and retrain KNN regressor
best_knn_regressor = knn_regressor_grid_search.best_estimator_
best_knn_regressor.fit(X_train, y_train)

# Assess performance of best models
best_knn_regressor_rmse = mean_squared_error(y_test, best_knn_regressor.predict(X_test), squared=False)

print("Best KNN Regressor RMSE:", best_knn_regressor_rmse)


/Users/mrburstein/anaconda3/lib/python3.11/site-packages/joblib/externals/loky/process_executor.py:700: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


Best KNN Regressor RMSE: 15611964.26232824


In [24]:
from sklearn.metrics import mean_absolute_error, silhouette_score, davies_bouldin_score, accuracy_score, precision_score, recall_score, f1_score, mean_squared_error, r2_score
# 4. Comparison
#4.4 Assess model

def assess_models(models, X_test, y_test):
    assessment_report = {}
    
    for model_name, model in models.items():
        assessment_report[model_name] = {}
        if 'clustering' in model_name.lower():
            y_pred = model.predict(X_test)
            # Clustering evaluation metrics
            y_pred = model.predict(X_test)
            accuracy = accuracy_score(y_test, y_pred)
            precision = precision_score(y_test, y_pred, average='weighted')
            recall = recall_score(y_test, y_pred, average='weighted')
            f1 = f1_score(y_test, y_pred, average='weighted')
            mae = mean_absolute_error(y_test, y_pred)
            mse = mean_squared_error(y_test, y_pred)
            inertia = model.inertia_
            db_index = davies_bouldin_score(X_test, model.labels_)
            r2 = r2_score(y_test, y_pred)


            assessment_report[model_name]['Inertia'] = inertia
            assessment_report[model_name]['Davies-Bouldin Index'] = db_index
            assessment_report[model_name]['R-squared'] = r2
            assessment_report[model_name]['mae'] = mae
            assessment_report[model_name]['Mean Squared Error'] = mse            
            assessment_report[model_name]['Accuracy'] = accuracy
            assessment_report[model_name]['Precision'] = precision
            assessment_report[model_name]['Recall'] = recall
            assessment_report[model_name]['F1-Score'] = f1
        elif 'best_knn_classifier' in model_name.lower():
            # KNN evaluation metrics
            y_pred = model.predict(X_test)
            accuracy = accuracy_score(y_test, y_pred)
            precision = precision_score(y_test, y_pred, average='weighted')
            recall = recall_score(y_test, y_pred, average='weighted')
            f1 = f1_score(y_test, y_pred, average='weighted')
            mae = mean_absolute_error(y_test, y_pred)
            mse = mean_squared_error(y_test, y_pred)
            r2 = r2_score(y_test, y_pred)

            assessment_report[model_name]['R-squared'] = r2
            assessment_report[model_name]['mae'] = mae
            assessment_report[model_name]['Mean Squared Error'] = mse            
            assessment_report[model_name]['Accuracy'] = accuracy
            assessment_report[model_name]['Precision'] = precision
            assessment_report[model_name]['Recall'] = recall
            assessment_report[model_name]['F1-Score'] = f1
        elif 'best_knn_r' in model_name.lower():
            # KNN evaluation metrics
            # KNN evaluation metrics
            y_pred = model.predict(X_test)
            mae = mean_absolute_error(y_test, y_pred)
            mse = mean_squared_error(y_test, y_pred)
            #inertia = model.inertia_
            #db_index = davies_bouldin_score(X_test, model.labels_)
            r2 = r2_score(y_test, y_pred)

            #assessment_report[model_name]['Inertia'] = inertia
            #assessment_report[model_name]['Davies-Bouldin Index'] = db_index
            assessment_report[model_name]['R-squared'] = r2
            assessment_report[model_name]['mae'] = mae
            assessment_report[model_name]['Mean Squared Error'] = mse            
            #assessment_report[model_name]['Accuracy'] = accuracy
            #assessment_report[model_name]['Precision'] = precision
            #assessment_report[model_name]['Recall'] = recall
            #assessment_report[model_name]['F1-Score'] = f1
        elif 'linear regression' in model_name.lower():
            # Linear Regression evaluation metrics
            y_pred = model.predict(X_test)
            # accuracy = accuracy_score(y_test, y_pred)
            # recall = recall_score(y_test, y_pred, average='weighted')
            # f1 = f1_score(y_test, y_pred, average='weighted')
            mae = mean_absolute_error(y_test, y_pred)
            mse = mean_squared_error(y_test, y_pred)
            # inertia = model.inertia_
            # db_index = davies_bouldin_score(X_test, model.labels_)
            r2 = r2_score(y_test, y_pred)


            # assessment_report[model_name]['Inertia'] = inertia
            # assessment_report[model_name]['Davies-Bouldin Index'] = db_index
            assessment_report[model_name]['R-squared'] = r2
            assessment_report[model_name]['mae'] = mae
            assessment_report[model_name]['Mean Squared Error'] = mse            
            # assessment_report[model_name]['Accuracy'] = accuracy
            #assessment_report[model_name]['Precision'] = precision
            # assessment_report[model_name]['Recall'] = recall
            # assessment_report[model_name]['F1-Score'] = f1
        elif 'best_ridge_model' in model_name.lower():
            # Linear Regression evaluation metrics
            y_pred = model.predict(X_test)
            # accuracy = accuracy_score(y_test, y_pred)
            # recall = recall_score(y_test, y_pred, average='weighted')
            # f1 = f1_score(y_test, y_pred, average='weighted')
            mae = mean_absolute_error(y_test, y_pred)
            mse = mean_squared_error(y_test, y_pred)
            # inertia = model.inertia_
            # db_index = davies_bouldin_score(X_test, model.labels_)
            r2 = r2_score(y_test, y_pred)


            # assessment_report[model_name]['Inertia'] = inertia
            # assessment_report[model_name]['Davies-Bouldin Index'] = db_index
            assessment_report[model_name]['R-squared'] = r2
            assessment_report[model_name]['mae'] = mae
            assessment_report[model_name]['Mean Squared Error'] = mse            
            # assessment_report[model_name]['Accuracy'] = accuracy
            #assessment_report[model_name]['Precision'] = precision
            # assessment_report[model_name]['Recall'] = recall
            # assessment_report[model_name]['F1-Score'] = f1
        elif 'kmeans' in model_name.lower():
            y_pred = model.predict(X_test)

            # KMeans evaluation metrics
            y_pred = model.predict(X_test)
            accuracy = accuracy_score(y_test, y_pred)
            precision = precision_score(y_test, y_pred, average='weighted')
            recall = recall_score(y_test, y_pred, average='weighted')
            f1 = f1_score(y_test, y_pred, average='weighted')
            mae = mean_absolute_error(y_test, y_pred)
            mse = mean_squared_error(y_test, y_pred)
            inertia = model.inertia_
            db_index = davies_bouldin_score(X_test, model.labels_)
            r2 = r2_score(y_test, y_pred)


            assessment_report[model_name]['Inertia'] = inertia
            assessment_report[model_name]['Davies-Bouldin Index'] = db_index
            assessment_report[model_name]['R-squared'] = r2
            assessment_report[model_name]['mae'] = mae
            assessment_report[model_name]['Mean Squared Error'] = mse            
            assessment_report[model_name]['Accuracy'] = accuracy
            assessment_report[model_name]['Precision'] = precision
            assessment_report[model_name]['Recall'] = recall
            assessment_report[model_name]['F1-Score'] = f1
    
    return assessment_report

# Example usage
models = {
    'KMeans': kmeans,
    'best_knn_classifier': best_knn_classifier,
    'best_knn_regressor': best_knn_regressor,
    'Linear Regression': linear_reg,
    'best_ridge_model': best_ridge_model
}

assessment_report = assess_models(models, X_train, y_train)
#assessment_report = assess_models(models, X_test, y_test)
print(assessment_report)


/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/mrburstein/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


{'KMeans': {'Inertia': 2.2647195510085203e+17, 'Davies-Bouldin Index': 0.5013513584361482, 'R-squared': 0.5450104122954191, 'mae': 54588.83642033733, 'Mean Squared Error': 49406193861692.79, 'Accuracy': 0.016154233628398084, 'Precision': 0.005673389765142149, 'Recall': 0.016154233628398084, 'F1-Score': 0.008338759298778692}, 'best_knn_classifier': {'R-squared': 0.9999999996050528, 'mae': 1.7713061026748331, 'Mean Squared Error': 42886.334442366126, 'Accuracy': 0.9997946631673849, 'Precision': 0.9997877241149636, 'Recall': 0.9997946631673849, 'F1-Score': 0.9997887852711658}, 'best_knn_regressor': {'R-squared': 1.0, 'mae': 0.0, 'Mean Squared Error': 0.0}, 'Linear Regression': {'R-squared': 8.775516617187673e-05, 'mae': 121392.74271680435, 'Mean Squared Error': 108577997272800.36}, 'best_ridge_model': {'R-squared': 8.775494915846949e-05, 'mae': 121372.75591408354, 'Mean Squared Error': 108577997296365.31}}


In [25]:
import matplotlib.pyplot as plt

# Convert assessment report to a DataFrame
df = pd.DataFrame.from_dict(assessment_report, orient='index')

# Plot the assessment report as a table
# plt.figure(figsize=(10, 6))
# plt.table(cellText=df.values,
#           colLabels=df.columns,
#           rowLabels=df.index,
#           loc='center')
# plt.axis('off')
# plt.title('Assessment Report')
# plt.show()

df.head()

Inertia  Davies-Bouldin Index  R-squared  \
KMeans               2.264720e+17              0.501351   0.545010   
best_knn_classifier           NaN                   NaN   1.000000   
best_knn_regressor            NaN                   NaN   1.000000   
Linear Regression             NaN                   NaN   0.000088   
best_ridge_model              NaN                   NaN   0.000088   

                               mae  Mean Squared Error  Accuracy  Precision  \
KMeans                54588.836420        4.940619e+13  0.016154   0.005673   
best_knn_classifier       1.771306        4.288633e+04  0.999795   0.999788   
best_knn_regressor        0.000000        0.000000e+00       NaN        NaN   
Linear Regression    121392.742717        1.085780e+14       NaN        NaN   
best_ridge_model     121372.755914        1.085780e+14       NaN        NaN   

                       Recall  F1-Score  
KMeans               0.016154  0.008339  
best_knn_classifier  0.999795  0.999789  
best_knn_regressor        NaN       NaN  
Linear Regression         NaN       NaN  
best_ridge_model          NaN       NaN

### Evaluation

With some modeling accomplished, we aim to reflect on what we identify as a high quality model and what we are able to learn from this.  We should review our business objective and explore how well we can provide meaningful insight on drivers of used car prices.  Your goal now is to distill your findings and determine whether the earlier phases need revisitation and adjustment or if you have information of value to bring back to your client.

## Evaluation Results

### Assessment of Data Mining Results

- **Model Performance**: The KNN Classifier model achieved an accuracy of 0.48985, while the KNN Regressor model has a very high Mean Squared Error of 8.626775e+13 and a low R-squared value of 0.205546. The Linear Regression model also has a very high Mean Squared Error of 1.085780e+14 and an extremely low R-squared value of 0.000088. The KMeans model's performance metrics are not available.
  
- **Summarization**: While the models evaluated did not meet the initial business objectives due to poor performance, there may still be value in proceeding with deployment to gather real-world feedback.

### Approved Models

- **Identified Models**: None of the models meet the selected criteria for approval due to unsatisfactory performance.

### Review Process

- **Thorough Review**: A thorough review of the data mining engagement was conducted.
  
- **Quality Assurance**: Quality assurance checks were performed on the correctness of the model-building process and attribute usage.

## Determination of Next Steps

### Potential Further Actions

- **Option 1**: Despite the subpar performance, consider finishing the project and moving to deployment to gather real-world feedback.
  
- **Option 2**: Initiating further iterations to improve models may still be necessary to refine performance.
  
- **Option 3**: Setting up new data mining projects for additional analyses may also be considered to explore alternative modeling approaches.

### Decision

- **Chosen Action**: The chosen action is to proceed with deployment despite the current models' poor performance.
  
- **Rationale**: While the models may not meet the desired performance metrics, deploying them can provide valuable insights and real-world feedback that may inform future iterations and improvements. This decision is based on the assessment results and acknowledges the potential benefits of gathering real-world data.


### Deployment

Now that we've settled on our models and findings, it is time to deliver the information to the client.  You should organize your work as a basic report that details your primary findings.  Keep in mind that your audience is a group of used car dealers interested in fine tuning their inventory.

## Deployment Plan

### Model Selection
Despite the unsatisfactory performance of the evaluated models, a decision has been made to proceed with deployment to gather real-world feedback.

### Preparation of Deployment Environment
1. **Infrastructure Setup**: Configure the necessary infrastructure for deploying the models, including servers, databases, and networking components.
2. **Environment Configuration**: Set up the deployment environment with required software dependencies, libraries, and frameworks.

### Model Export
1. **Export Models**: Export the selected models from the development environment to the deployment environment.
2. **Compatibility Check**: Ensure that the deployed models are compatible with the deployment environment and can be executed without errors.

### Integration
1. **Integration with Existing Systems**: Integrate the deployed models into existing systems or applications where they will be used for making predictions or decisions.
2. **Testing**: Conduct thorough testing to ensure seamless integration and functionality within the production environment.

### Documentation
1. **Deployment Documentation**: Document the deployment process, including steps taken, configurations made, and any troubleshooting procedures.
2. **User Guide**: Prepare a user guide or documentation to assist end-users in understanding and utilizing the deployed models effectively.

## Monitoring and Maintenance Plan

### Performance Monitoring
1. **Real-time Monitoring**: Implement mechanisms to monitor the performance of deployed models in real-time, including accuracy, response time, and resource utilization.
2. **Alerting System**: Set up an alerting system to notify administrators of any performance degradation or errors encountered during model execution.

### Error Handling
1. **Error Logging**: Configure error logging to capture and log any errors or exceptions that occur during model deployment or execution.
2. **Error Resolution**: Establish protocols for handling and resolving errors promptly to minimize disruptions to business operations.

### Regular Maintenance
1. **Scheduled Maintenance Tasks**: Schedule regular maintenance tasks, such as model retraining, software updates, and data validation, to ensure continued effectiveness and accuracy of the deployed models.
2. **Data Quality Checks**: Implement automated data quality checks to monitor the quality and integrity of input data used by the models.

### Feedback Loop
1. **Feedback Collection**: Establish a feedback loop to gather feedback from end-users and stakeholders regarding the performance and usability of the deployed models.
2. **Continuous Improvement**: Use feedback gathered to identify areas for improvement and iterate on the deployed models to enhance their performance and functionality over time.

## Decision and Rationale
The decision to proceed with deployment despite the poor performance of the models is based on the potential benefits of gathering real-world feedback. By deploying the models, valuable insights can be gained that may inform future iterations and improvements, ultimately leading to better-performing models aligned with the business objectives.
